### PR TITLE
[ML] Add hyperparameter model metadata

### PR DESCRIPTION
**Pull request by valeriy42**
This adds `hyperparameters` section to the stored model metadata document. Here is an example
```json
"hyperparameters": [
        {
            "name": "downsample_factor",
            "value": 0.35355339059327386,
            "supplied": false,
            "absolute_importance": 0.123,
            "relative_importance": 0.9123
        },
        {
            "name": "alpha",
            "value": 0.010766199582740555,
            "supplied": false,
            "absolute_importance": 0.0123,
            "relative_importance": 0.0321
        },
        {
            "name": "gamma",
            "value": 0.1,
            "supplied": true,
            "absolute_importance": 0.0,
            "relative_importance": 0.0
        }
    ]
```

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java [hunk 1]
```diff
@@ -24,23 +25,29 @@
...
+
     }
 
     private final List<TotalFeatureImportance> featureImportances;
     private final FeatureImportanceBaseline featureImportanceBaseline;
+    private final List<Hyperparameters> hyperparameters;
 
-    public ModelMetadata(List<TotalFeatureImportance> featureImportances, FeatureImportanceBaseline featureImportanceBaseline) {
+    public ModelMetadata(List<TotalFeatureImportance> featureImportances, FeatureImportanceBaseline featureImportanceBaseline, 
+    List<Hyperparameters> hyperparameters) {
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java:47]:
```suggestion
    public ModelMetadata(List<TotalFeatureImportance> featureImportances, 
        FeatureImportanceBaseline featureImportanceBaseline, 
        List<Hyperparameters> hyperparameters) {
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java [hunk 2]
```diff
@@ -72,6 +84,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         if (featureImportanceBaseline != null) {
             builder.field(FEATURE_IMPORTANCE_BASELINE.getPreferredName(), featureImportanceBaseline);
         }
+        builder.field(HYPERPARAMETERS.getPreferredName(), hyperparameters);
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java:87]:
```suggestion
        if (hyperparameters != null) {
            builder.field(HYPERPARAMETERS.getPreferredName(), hyperparameters);
        }
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadata.java [hunk 3]
```diff
@@ -38,14 +39,18 @@
...
+            a -> new TrainedModelMetadata((String)a[0], (List<TotalFeatureImportance>)a[1], (FeatureImportanceBaseline)a[2], 
+                                          (List<Hyperparameters>)a[3]));
         parser.declareString(ConstructingObjectParser.constructorArg(), MODEL_ID);
         parser.declareObjectArray(ConstructingObjectParser.constructorArg(),
             ignoreUnknownFields ? TotalFeatureImportance.LENIENT_PARSER : TotalFeatureImportance.STRICT_PARSER,
             TOTAL_FEATURE_IMPORTANCE);
         parser.declareObject(ConstructingObjectParser.optionalConstructorArg(),
             ignoreUnknownFields ? FeatureImportanceBaseline.LENIENT_PARSER : FeatureImportanceBaseline.STRICT_PARSER,
             FEATURE_IMPORTANCE_BASELINE);
+        parser.declareObjectArray(ConstructingObjectParser.constructorArg(),
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadata.java:51]: Fixed

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadataTests.java [hunk 4]
```diff
@@ -24,7 +24,8 @@ public static TrainedModelMetadata randomInstance() {
         return new TrainedModelMetadata(
             randomAlphaOfLength(10),
             Stream.generate(TotalFeatureImportanceTests::randomInstance).limit(randomIntBetween(1, 10)).collect(Collectors.toList()),
-            randomBoolean() ? null : FeatureImportanceBaselineTests.randomInstance());
+            randomBoolean() ? null : FeatureImportanceBaselineTests.randomInstance(), 
+            Stream.generate(HyperparametersTests::randomInstance).limit(randomIntBetween(1, 10)).collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadataTests.java:28]: fixed

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java [hunk 5]
```diff
@@ -24,23 +25,29 @@
...
     @SuppressWarnings("unchecked")
     public static final ConstructingObjectParser<ModelMetadata, Void> PARSER = new ConstructingObjectParser<>(
         "trained_model_metadata",
-        a -> new ModelMetadata((List<TotalFeatureImportance>) a[0], (FeatureImportanceBaseline) a[1]));
+        a -> new ModelMetadata((List<TotalFeatureImportance>) a[0], (FeatureImportanceBaseline) a[1], (List<Hyperparameters>) a[2]));
 
     static {
         PARSER.declareObjectArray(constructorArg(), TotalFeatureImportance.STRICT_PARSER, TOTAL_FEATURE_IMPORTANCE);
         PARSER.declareObject(optionalConstructorArg(), FeatureImportanceBaseline.STRICT_PARSER, FEATURE_IMPORTANCE_BASELINE);
+        PARSER.declareObjectArray(constructorArg(), Hyperparameters.STRICT_PARSER, HYPERPARAMETERS);
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java:38]: Fixed

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java (hunk 2 above)
Comment (valeriy42) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java:87]: fixed

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java (hunk 1 above)
Comment (valeriy42) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/results/ModelMetadata.java:47]: I don't see the change here. Is this a formatting issue?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadata.java [hunk 6]
```diff
@@ -63,20 +68,25 @@ public static String modelId(String docId) {
...
     public TrainedModelMetadata(String modelId,
                                 List<TotalFeatureImportance> totalFeatureImportances,
-                                FeatureImportanceBaseline featureImportanceBaselines) {
+                                FeatureImportanceBaseline featureImportanceBaselines,
+                                List<Hyperparameters> hyperparameters) {
         this.modelId = ExceptionsHelper.requireNonNull(modelId, MODEL_ID);
         this.totalFeatureImportances = Collections.unmodifiableList(totalFeatureImportances);
         this.featureImportanceBaselines = featureImportanceBaselines;
+        this.hyperparameters = Collections.unmodifiableList(hyperparameters);
+
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadata.java:89]: fixed

**Comment by valeriy42** [comment]
Thank you for your review @benwtrent . I applied your changes. It would be great if you could take another look.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        this.relativeImportance = in.readDouble();
+        this.supplied = in.readBoolean();
+    }
+
+    Hyperparameters(String hyperparameterName, Double value, Double absoluteImportance, Double relativeImportance, Boolean supplied) {
+        this.hyperparameterName = hyperparameterName;
+        this.value = value;
+        this.absoluteImportance = absoluteImportance;
+        this.relativeImportance = relativeImportance;
+        this.supplied = supplied;
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java:73]: If the `Double`s and the `Boolean`  are truly not nullable, they should be `double` and `boolean` (unboxed, not nullable). 

If they are nullable, the parser needs to treat them as optional and the serialization needs to as well

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadata.java [hunk 8]
```diff
@@ -95,26 +105,33 @@ public FeatureImportanceBaseline getFeatureImportanceBaselines() {
...
+        return Objects.hash(totalFeatureImportances, featureImportanceBaselines, hyperparameters, modelId);
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeString(modelId);
         out.writeList(totalFeatureImportances);
         out.writeOptionalWriteable(featureImportanceBaselines);
+        out.writeList(hyperparameters);
+
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadata.java:134]:
```suggestion
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadata.java [hunk 9]
```diff
@@ -63,20 +68,25 @@ public static String modelId(String docId) {
...
     public TrainedModelMetadata(String modelId,
                                 List<TotalFeatureImportance> totalFeatureImportances,
-                                FeatureImportanceBaseline featureImportanceBaselines) {
+                                FeatureImportanceBaseline featureImportanceBaselines,
+                                List<Hyperparameters> hyperparameters) {
         this.modelId = ExceptionsHelper.requireNonNull(modelId, MODEL_ID);
         this.totalFeatureImportances = Collections.unmodifiableList(totalFeatureImportances);
         this.featureImportanceBaselines = featureImportanceBaselines;
+        this.hyperparameters =  hyperparameters == null ? Collections.emptyList() : Collections.unmodifiableList(hyperparameters);
+
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/TrainedModelMetadata.java:89]:
```suggestion
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        this.value = in.readDouble();
+        this.supplied = in.readBoolean();
+        if (this.supplied == false) {
+            this.absoluteImportance = in.readDouble();
+            this.relativeImportance = in.readDouble();
+        }
+        else {
+            this.absoluteImportance = null;
+            this.relativeImportance = null;
+        }
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java:72]:
```suggestion
        this.absoluteImportance = in.readOptionalDouble();
        this.relativeImportance = in.readOptionalDouble();
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        this.value = value;
+        this.supplied = supplied;
+        if (this.supplied == false) {
+            this.absoluteImportance = ExceptionsHelper.requireNonNull(absoluteImportance, ABSOLUTE_IMPORTANCE.getPreferredName());
+            this.relativeImportance = ExceptionsHelper.requireNonNull(relativeImportance, RELATIVE_IMPORTANCE.getPreferredName());
+        }
+        else {
+            this.absoluteImportance = null;
+            this.relativeImportance = null;
+        }
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java:86]:
```suggestion
        this.absoluteImportance = absoluteImportance;
        this.relativeImportance = relativeImportance;
        if (this.supplied == false) {
            ExceptionsHelper.requireNonNull(absoluteImportance, ABSOLUTE_IMPORTANCE.getPreferredName());
            ExceptionsHelper.requireNonNull(relativeImportance, RELATIVE_IMPORTANCE.getPreferredName());
        }
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(hyperparameterName);
+        out.writeDouble(value);
+        out.writeBoolean(supplied);
+        if (supplied == false) {
+            out.writeOptionalDouble(absoluteImportance);
+            out.writeOptionalDouble(relativeImportance);
+        }
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java:97]:
```suggestion
        out.writeOptionalDouble(absoluteImportance);
        out.writeOptionalDouble(relativeImportance);
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        map.put(HYPERPARAMETER_NAME.getPreferredName(), hyperparameterName);
+        map.put(VALUE.getPreferredName(), value);
+        if (supplied == false) {
+            if (absoluteImportance != null) {
+                map.put(ABSOLUTE_IMPORTANCE.getPreferredName(), absoluteImportance);
+            }
+            if (relativeImportance != null) {
+                map.put(RELATIVE_IMPORTANCE.getPreferredName(), relativeImportance);
+            }
+        }
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/Hyperparameters.java:129]:
```suggestion
        if (absoluteImportance != null) {
            map.put(ABSOLUTE_IMPORTANCE.getPreferredName(), absoluteImportance);
        }
        if (relativeImportance != null) {
            map.put(RELATIVE_IMPORTANCE.getPreferredName(), relativeImportance);
        }
```

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/HyperparametersTests.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+import java.io.IOException;
+
+
+public class HyperparametersTests extends AbstractBWCSerializationTestCase<Hyperparameters> {
+
+    private boolean lenient;
+
+    @SuppressWarnings("unchecked")
+    public static Hyperparameters randomInstance() {
+        return new Hyperparameters(
```
Comment (benwtrent) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/HyperparametersTests.java:23]:
```suggestion
        boolean supplied = randomBoolean();
        return new Hyperparameters(
```

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/HyperparametersTests.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+            randomAlphaOfLength(10),
+            randomDoubleBetween(0.0, 1.0, true),
+            randomDoubleBetween(0.0, 100.0, true),
+            randomDoubleBetween(0.0, 1.0, true),
+            randomBoolean());
```
Comment (benwtrent) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/metadata/HyperparametersTests.java:28]:
```suggestion
            // If supplied, the importance values are possibly nullable
            supplied && randomBoolean() ? null : randomDoubleBetween(0.0, 100.0, true),
            supplied && randomBoolean() ? null : randomDoubleBetween(0.0, 1.0, true),
            supplied);
```

**Review by benwtrent — APPROVED** [review]

**Comment by valeriy42** [comment]
retest

**Comment by valeriy42** [comment]
recheck

File: x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/core/ml/inference_index_template.json [hunk 16]
```diff
@@ -127,6 +127,27 @@
...
+            },
+            "absolute_importance": {
+              "type": "double"
+            },
+            "relative_importance": {
+              "type": "double"
+            },
+            "supplied": {
+              "type": "boolean"
+            },
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/resources/org/elasticsearch/xpack/core/ml/inference_index_template.json:149]: ^ dangling comma

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**